pkg/aggregator/internal/limiter/limiter.go

```diff
@@ -0,0 +1,165 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package limiter
+
+import (
+	"strings"
+
+	"github.com/DataDog/datadog-agent/pkg/metrics"
+	"github.com/DataDog/datadog-agent/pkg/tagset"
+)
+
+type entry struct {
+	current  int // number of contexts currently in aggregator
+	rejected int // number of rejected samples
+	tags     []string
+}
+
+// Limiter tracks number of contexts based on origin detection metrics
+// and rejects samples if the number goes over the limit.
+//
+// Not thread safe.
+type Limiter struct {
+	key   string
+	tags  []string
+	limit int
+	usage map[string]*entry
+}
+
+// New returns a new instance of limiter.
+//
+// If limit is zero or less the limiter is disabled.
+func New(limit int, key string, tags []string) *Limiter {
+	if limit <= 0 {
+		return nil
+	}
+
+	if !strings.HasSuffix(key, ":") {
+		key += ":"
+	}
+
+	hasKey := false
+	tags = append([]string{}, tags...)
+	for i := range tags {
+		if !strings.HasSuffix(tags[i], ":") {
+			tags[i] += ":"
+		}
+		hasKey = hasKey || key == tags[i]
+	}
+
+	if !hasKey {
+		tags = append(tags, key)
+	}
+
+	return &Limiter{
+		key:   key,
+		tags:  tags,
+		limit: limit,
+		usage: map[string]*entry{},
+	}
+}
+
+func (l *Limiter) identify(tags []string) string {
+	for _, t := range tags {
+		if strings.HasPrefix(t, l.key) {
+			return t
+		}
+	}
+	return ""
+}
+
+func (l *Limiter) extractTags(src []string) []string {
+	dst := make([]string, 0, len(l.tags))
+
+	for _, t := range src {
+		for _, p := range l.tags {
+			if strings.HasPrefix(t, p) {
+				dst = append(dst, t)
+			}
+		}
+	}
+
+	return dst
+}
+
+// Track is called for each new context. Returns true if the sample should be accepted, false
+// otherwise.
+func (l *Limiter) Track(tags []string) bool {
+	if l == nil {
+		return true
+	}
+
+	id := l.identify(tags)
+
+	e := l.usage[id]
+	if e == nil {
+		e = &entry{
+			tags: l.extractTags(tags),
+		}
+		l.usage[id] = e
+	}
```

Comment:
In the case where `key` is `pod_name` and `l.tags` is `container_id`, when we start tracking the context, we'll only capture the first `container_id` tag that is sent, so if multiple containers within a pod are emitting metrics, the telemetry tags will only list one `container_id` despite there being multiple containers that contributed to the quota.

Should we document this limitation/assumption somewhere? Or extract the tags for subsequent calls to `Track` as well? (I believe we support tags with multiple values, but also would get tricky with expiring contexts)